speclang_modern.tex

```diff
@@ -936,7 +936,8 @@ \subsection{Built-in constructs %
   predicate or term \lstinline|e| in the pre-state of the function.
 \item \result\indexttbs{result} denotes the returned value of the function.
 \end{itemize}
-\lstinline|\old(e)| and \result{} can be used only in \ensures{} clauses, since
+\lstinline|\old(e)| and \result{} can be used only in \ensures{}, \assigns{},
+\allocates{} and \frees{} clauses, since
```

Comment:
> The only difference between `allocates` and `frees` is that sets `Pi` are evaluated in the pre-state, and sets `Qi` are evaluated in the post-state.

Reply:
Anyway, there is something special for `\result`, since it is a default output and  is allowed into `assigns` clauses, only as a *valid output location* for being able to gives the its `\from` dependencies. Its use is forbiden at all other places in theses clauses (i.e. `assigns tab[\result];` is not allowed, nor `assigns x \from \result, tab[\result];`). So, a restriction has to be detailed in that use.

Reply:
Is that so? I could imagine some use for `assigns tab[\at(\result,Post)];` (possibly with expliciting the `\at(.,Post)` indeed), e.g. for

```c
int t[42];

/*@ assigns \result \from x;
       assigns t[\at(\result,Post)] \from x;
*/
int foo(int x) {
  int idx = complex_computation(x);
  t[idx] = 1;
  return idx;
}       
```

For me, the real constraint for being able to use `\result` is that we are in the post-state of _a function_ (this does not apply to statement contract, but note that mentioning `\result` in the `ensures` clause of such contract does not really make sense either. Only `returns` can have it), either implicitely (`ensures` clause of a function contract), or explicitely (`\at(e,Post)`), without another `\at()` getting in the way.

Reply:
I agree. Sorry, I had forgotten that point.

Reply:
Just to be sure, as the discussion seems to be going elsewhere: my point was that `\result` is meaningless in `frees` clauses.

Reply:
@yakobowski  I was about to make an untimely comment about the fact that `allocates a;` is equivalent to `frees \at(a,Post)` 😵, but after re-reading the section, I have to admit that it does not say anything about the built-in labels that can be used in the respective clauses.

Reply:
At the very least, all clauses should not be coalesced together in the text, since `frees` is by default in the pre-state, and the other in the post state.

But I'm not convinced that `Post` should be valid in `frees` clauses, except to make things even more complicated for everyone. You should be able to express what you are freeing in terms intelligible to your caller, or you have a big problem.

Reply:
Well, there are cases where `Post` could be handy in a `frees` clause. For instance, take the function below (slight variation over `foo` above):

```c
struct { int allocated;  unsigned char* data; } t[42];
/*@ frees t[\at(\result,Post)].data;
       assigns \result \from x;
       assigns t[\at(\result,Post)] \from x, t[\at(\result,Post)];
*/
int bar(int x) {
  int res = complex_computation(x);
  if (t[res].allocated) {
     free(t[res].data);
     t[res].allocated = 0;
  }
  return res;
}
```

Reply:
Agree, also complex_computation() can be not only complex, but nondeterministic as well.